### DISCUSSION TITLE
What is the role of SupportedFeatures.java?

### DISCUSSION THREAD

**Discussion by turkeylurkey** [redhat-developer/lsp4ij#271] [Q&A] [answered]
When code in SupportedFeatures.java enables completion with resolve does that describe the client or the server? If the server does not support this capability is it disabled in the local server capabilities configuration object? What requirements are on the client when completion with resolve is enabled?

**Comment by angelozerr** [comment, marked as answer]
At first this class name comes from LSP4E. I think we should rename to ClientSupportedFeatures or à better name.

This class contains all capabilities that LSP4IJ can support and those capabilities are send to language servers which must be started on initialize step.

The language server can use those capability to adapt the behavior of features. 

For instance MP ls returns completion with documentation for ech items if the lsp client cannot support resolve completion.

With LSP4IJ the MP ls will return completion without documentation since it can support resolve completion.

When lsp4ij need to show documentation for the delected item it will consume the completion redolve of mp ls because mp ls can support resolve completion

**Reply by angelozerr** [reply]
> When code in SupportedFeatures.java enables completion with resolve does that describe the client or the server? 

The client.

> If the server does not support this capability is it disabled in the local server capabilities configuration object? 

The server capabilities is different from client capabilities. This capabilities comes from theblanguage server himself.

I would say Yes and lsp4ij will not consume resolve completion.

> What requirements are on the client when completion with resolve is enabled?

Lsp4ij can resolve detail documrntation and additionalTextEdits if those fields are not filled

**Comment by angelozerr** [comment]
@turkeylurkey can we close this discussion or do you need more informations?

**Comment by turkeylurkey** [comment]
Thank you for the explanation. Since the client references the server capabilities object locally I wanted to make sure the local object is correct. I agree that a rename of the class would help.

**Reply by angelozerr** [reply]
You are welcome. Please review https://github.com/redhat-developer/lsp4ij/pull/288
